test-suite/handwritten-src/objc/tests/DBAsyncTest.m

```diff
@@ -0,0 +1,79 @@
+#import <Foundation/Foundation.h>
+#import <XCTest/XCTest.h>
+#import <DJFuture.h>
+#import "DBTestHelpers.h"
+#import "DBAsyncInterface.h"
+
+@interface AsyncInterfaceImpl : NSObject <DBAsyncInterface>
+@end
+
+@implementation AsyncInterfaceImpl
+- (DJFuture<NSString *> *)futureRoundtrip:(DJFuture<NSNumber *> *)f {
+    return [f then:^id(DJFuture<NSNumber*>* i) { return [[i get] stringValue]; }];
+}
+@end
+
+@interface DBAsyncTests : XCTestCase
+@end
+
+@implementation DBAsyncTests
+
+- (void)setUp
+{
+    [super setUp];
+}
+
+- (void)tearDown
+{
+    [super tearDown];
+}
+
+- (void)testConsumeNativeFuture {
+    DJFuture<NSNumber*>* f1 = [DBTestHelpers getAsyncResult];
+    DJFuture<NSString*>* f2 =[f1 then:^id(DJFuture<NSNumber*>* res) {
+            return [[res get] stringValue];
+        }];
+    DJFuture<NSNumber*>* f3 = [f2 then:^id(DJFuture<NSString*>* res) {
+            return @([[res get] integerValue]);
+        }];
+    NSNumber* i = [f3 get];
+    XCTAssertEqual([i integerValue], 42);
```

Comment:
It took me 5 minutes for me to figure out what's with this magic number `42`.
It would be nice to see it in this test. e.g.,
```ObjC
DJFuture<NSNumber *> *f1 = [DBTestHelpers getAsyncResultWithValue:@42];
```

Reply:
good idea

Reply:
I tried that but it's not straightforward because of this code:
```
#if defined(__EMSCRIPTEN__)
    emscripten_async_call([] (void* context) {
        auto* p = reinterpret_cast<snapchat::djinni::Promise<int32_t>*>(context);
        p->setValue(42);
        delete p;
    }, p, 10/*ms*/);
#else
```
Here `emscripten_async_call` must take a function pointer.  If I capture anything into this lambda here, it will not be convertible to a plain function pointer.

Reply:
I'm not sure if I'm really following you on this. Can't we just do this?

```cpp
snapchat::djinni::Future<int32_t> TestHelpers::get_async_result(int32_t value) {
     // ...

 #if defined(__EMSCRIPTEN__)
     emscripten_async_call([value] (void* context) {
         auto* p = reinterpret_cast<snapchat::djinni::Promise<int32_t>*>(context);
         p->setValue(value);
         delete p;
     }, p, 10/*ms*/);
 #else
```

Reply:
No, this will not work.  A lambda like this `[] { ... }` (without captures) can be converted to a plain C function pointer.  While a lambda like this `[value] { ... }` cannot be converted to a C function pointer (it is its own special lambda type).  The function `emscripten_async_call` expects a function pointer,  passing it a lambda will produce a compilation error.

We can, of course, allocate a context struct on the heap and put the value there, but that is a  bit too much trouble than its worth.

Reply:
Fair enough!